source/common/upstream/upstream_impl.cc

```diff
@@ -734,6 +739,15 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(const HostVector& new_hosts,
           }
         }
 
+        // Did metadata change?
+        if (!metadataEq(host->metadata(), (*i)->metadata())) {
+          (*i)->metadata(host->metadata());
+          (*i)->canary(host->canary());
+
+          // If metadata changed, we need to rebuild. See github issue #3810.
+          health_changed = true;
```

Comment:
Probably should rename this var to reflect that it's tracking changes to more than just health.